UNK tokenizer class

ACE07-Sev opened the pull request:
closes #84.

To use, we have to :

words= ['photo', 'take']
rewriter = Rewriter([ UNKRewriteRule(words = words)])
tokenized_diagram= rewriter(diagram)

![image](https://github.com/CQCL/lambeq/assets/73689800/9739951e-2113-4686-b19b-979803d68d11)

With words = ['photo', 'take'] we have :
![image](https://github.com/CQCL/lambeq/assets/73689800/d1503485-abc1-4537-9f93-0c1dd8978d16)

ACE07-Sev commented:
@le-big-mac dear, I made the handle module you asked in the other PR. Please kindly review to see if it's what you requested.

le-big-mac commented on lambeq/rewrite/base.py:
```diff
@@ -393,3 +419,34 @@ def _ar(self, box: Box) -> Diagram:
     def _ob(self, ob: Ty) -> Ty:
         return ob
+
+def HandleUnknownWords(diagrams: Container[Diagram], training_data: Container[str] | None = None, test_data: Container[str] | None = None, train_status: bool | None = True, min_freq: int | None = 0) -> Container[Diagram]:
```
It would make more sense for `HandleUnknownWords` to be a class that has separate methods for training and test data, and can possibly store the vocabulary calculated during train time, this current method is fairly convoluted.

le-big-mac reviewed (commented):
Hi Ace, I think there is good initial progress here, but I have left some comments that need addressing. Of significant importance too is that there need to be tests added to this PR, as otherwise I cannot be sure that the code is always going to do the right thing.

ACE07-Sev commented:
Greetings Charlie,

Hope you are well. Thank you so much for the kind review and feedback! Ooh yes the syntax error I forgot to add the [] after I resolved it in my notebook. I also removed the redundant match check condition for rewrite. I will now focus on the handler.

ACE07-Sev commented:
@le-big-mac ?

ACE07-Sev commented:
@dimkart dear, may I ask if you could kindly guide me on how I can run the checks/tests here?

dimkart commented:
@ACE07-Sev Hi, only maintainers can run tests (will do it in a sec). @le-big-mac will respond to other questions you have soon.

dimkart commented:
@ACE07-Sev Of course you can run the tests locally though.

ACE07-Sev commented:
@dimkart dear, how can I run the tests locally? I am testing it like a novice I think hehe. I just use a notebook to make sure it works like it's supposed to. Hence the diagrams I sent before.

How can I make sure it passes the tests you run online locally?

ACE07-Sev commented:
@le-big-mac dear, thank you so much for the kind explanation! I added test for unkrewriterule (that's what I assume you meant as tests), and fixed the handler based on your description.

dimkart commented on lambeq/rewrite/base.py:
```diff
@@ -393,3 +422,96 @@ def _ar(self, box: Box) -> Diagram:
+        
+        return diagrams
+    
+    def handle_test(self) -> Container[Diagram]:
```
The diagrams are not passed here, so line 502 will fail below.

ACE07-Sev replied:
I see! I'll fix it right away!

dimkart commented:
> how can I run the tests locally?

Just run `pytest` from within your local lambeq repo. You might have to exclude a couple of files, see the `build_test` action for which ones. Alternatively, just run `pytest <specific_test_file.py>`.

ACE07-Sev commented:
What is the new error exactly? I don't understand the doc too long error.

dimkart commented:
@ACE07-Sev It would be great if you could make the tests work by tomorrow 13/6, which is the last day of the hackathon.

ACE07-Sev commented:
Understood, will do.

ACE07-Sev commented:
May I later also add a bit more to the HandleUnknownWords class? Within 2 hours or so. Just adding a bit more into the features of the class, i.e., handling single diagram VS list of diagrams automatically.

le-big-mac commented:
I think you have until the end of the 13th of June [anywhere on earth](https://time.is/Anywhere_on_Earth) to add to this PR, so please go ahead.

ACE07-Sev commented:
Greetings, 

May I ask if I could get a test run here? I can't run my tests in my repo anymore because of this :

GitHub Actions is currently disabled for your account. Please reach out to [GitHub Support](https://support.github.com/contact?tags=dotcom-actions) for assistance.

ACE07-Sev commented:
I was able to only check for all except for the build_and_test when it was running test notebooks.

ACE07-Sev commented:
Can I get a test to be run here please?

dimkart commented:
> Can I get a test to be run here please?

Github doesn't let us, since your last tests appear as "queued" for the last 17 hours; when we try to cancel we get the message "Failed to cancel workflow". Perhaps you can try to see if there is anything you can do from your side?

Thommy257 commented:
Hi @ACE07-Sev, I've sent an email to the github support. I'll let you know when this is fixed.

ACE07-Sev commented:
Greetings dear @dimkart and @Thommy257 , 

Hope you are well. Thank you so much for the heads up. I cancelled the queued workflow from my side. However, I can't run anything so I hope at least cancelling the other one helps.

Thank you so much dear Thommy, I'll be eagerly looking forward to the update!

dimkart commented:
> Greetings dear @dimkart and @Thommy257 ,
> 
> Hope you are well. Thank you so much for the heads up. I cancelled the queued workflow from my side. However, I can't run anything so I hope at least cancelling the other one helps.
> 
> Thank you so much dear Thommy, I'll be eagerly looking forward to the update!

@ACE07-Sev There is another one currently queued, can you try to cancel that one as well?

ACE07-Sev commented:
Of course, one moment...

ACE07-Sev commented:
Thommy dear, I don't have any actions in queue, nor in progress, or neutral, or pending.

ACE07-Sev commented:
![image](https://github.com/CQCL/lambeq/assets/73689800/d597f9c3-7a24-4f14-85a0-63d529f7e069)

ACE07-Sev commented:
If it's causing hindrance, I can ask a friend to fork my repo and make a PR here on my behalf, or I can make a new account, and fork it and make a new PR.

dimkart reviewed (approved):
Thanks for your work! You will be assigned to the issue together with another use, but the PRs will be merged later.
